system/core/pages.php

```diff
@@ -47,17 +47,57 @@ abstract class HyphaSystemPage extends HyphaPage {
 		abstract class for a page backed by datatype and xml file.
 	*/
 	abstract class HyphaDatatypePage extends HyphaPage {
+		const INDEX_TABLE_COLUMNS_TITLE = 'title';
+		const INDEX_TABLE_COLUMNS_AUTHOR = 'author';
+		const INDEX_TABLE_COLUMNS_DATE = 'date';
+		const INDEX_TABLE_COLUMNS_STATUS = 'status';
+
 		public $pageListNode, $language, $pagename, $privateFlag;
 
 		function __construct(HyphaDomElement $node, RequestContext $O_O) {
 			parent::__construct($O_O);
 			$this->replacePageListNode($node);
 		}
 
+		/**
+		 * @return string
+		 */
 		public static function getDatatypeName() {
 			return str_replace('_', ' ', get_called_class());
 		}
 
+		/**
+		 * @return array
+		 */
+		public static function getIndexTableColumns() {
+			return [
+				__(self::INDEX_TABLE_COLUMNS_TITLE),
+				__(self::INDEX_TABLE_COLUMNS_AUTHOR),
+				__(self::INDEX_TABLE_COLUMNS_DATE),
+			];
+		}
+
+		/**
+		 * @return array
+		 */
+		public function getIndexData() {
+			$id = $this->pageListNode->getAttribute('id');
+			$date = $this->getSortDateTime();
+			$dataMtx = [
+				__(self::INDEX_TABLE_COLUMNS_TITLE) => [
+					'class' => 'type_'.get_class($this).' '.($this->privateFlag ? 'is-private' : 'is-public'),
+					'sort' => preg_replace("/[^A-Za-z0-9]/", '', $this->getTitle()).'_'.$id,
+					'value' => '<a href="'.$this->language.'/'.$this->pagename.'">'.$this->getTitle().'</a>',
+				],
+				__(self::INDEX_TABLE_COLUMNS_AUTHOR) => $this->getAuthor(),
+				__(self::INDEX_TABLE_COLUMNS_DATE) => [
+					'sort' => $date ? $date->format('YmdHis') : '',
+					'value' => $date ? $date->format('Y-m-d') : '',
+				],
+			];
+			return array_intersect_key($dataMtx, array_fill_keys(self::getIndexTableColumns(), ''));
```

Comment:
What is the point of this array_intersect? This seems to drop any keys from `$dataMtx` that do not appear in `getIndexTableColumns()`, but in the code I can see that this does not happen? Or is the point that subclasses can *delete* columns from their superclasses by removing them from `getIndexTableColumns (by overriding it)`? If so, I'd suggest that those superclasses can also just delete the columns from `getIndexData` (by overriding that as well), or maybe index page should globally just do this intersect (rather than each datatype having to do it individually)?